pageserver/src/tenant/timeline.rs

```diff
@@ -1135,6 +1158,162 @@ impl Timeline {
+        let now = SystemTime::now();
+        let remote_client = match self.remote_client.as_ref() {
+            None => {
+                error!("no remote storage configured but eviction requires that");
+                // check again in 1 hour; XXX config watch mechanism
+                tokio::select! {
+                    _ = tokio::time::sleep(Duration::from_secs(3600)) => {},
+                    _ = task_mgr::shutdown_watcher() => {},
+                }
+                return;
+            }
+            Some(c) => c,
+        };
+        #[allow(dead_code)]
+        #[derive(Debug, Default)]
+        struct EvictionStats {
+            not_considered_due_to_clock_skew: usize,
+            candidates: usize,
+            evicted: usize,
+            errors: usize,
+            not_evictable: usize,
+        }
+        let mut stats = EvictionStats::default();
+        // Gather layers for eviction.
+        // NB: all the checks can be invalidated as soon as we release the layer map lock.
+        // We don't want to hold the layer map lock during eviction.
+        // So, we just need to deal with this.
+        let candidates: Vec<Arc<dyn PersistentLayer>> = {
+            let layers = self.layers.read().unwrap();
+            let mut candidates = Vec::new();
+            for hist_layer in layers.iter_historic_layers() {
+                if hist_layer.is_remote_layer() {
+                    continue;
+                }
+                let last_activity_ts = match hist_layer
+                    .access_stats()
+                    .most_recent_access_or_residence_event()
+                {
+                    Either::Left(mra) => mra.when,
+                    Either::Right(re) => re.timestamp,
+                };
+                let no_activity_for = match now.duration_since(last_activity_ts) {
+                    Ok(d) => d,
+                    Err(_e) => {
+                        // NB: don't log the error. If there are many layers and the system clock
+                        // is skewed, we'd be flooding the log.
+                        stats.not_considered_due_to_clock_skew += 1;
+                        continue;
+                    }
+                };
+                if no_activity_for > p.threshold {
+                    candidates.push(hist_layer)
+                }
+            }
+            candidates
+        };
+        stats.candidates = candidates.len();
+        for l in candidates {
+            // XXX would be good to batch the evictions.
+            // The way it is now, we have to rebuild the layer map after every eviction.
+            // Idea for this:
+            // - a layer should have an 'evicting' state. We set it here.
+            // - If GC / Compaction find such a layer, they skip it or otherwise
+            //   guarantee to not remove it from the layer map until 'evicting' state is over.
```

Comment:
Why this instead of the lock which is currently being used?

Reply:
Actually I think my performance conerns are unfounded with `LayerMap::replace`, aren't they?

Anyways, batch API seems useful. WDYT: https://github.com/neondatabase/neon/pull/3554 ?

Reply:
Yeah the batching impl looked exactly like I was thinking, forgot to add any positive comments on the PR before you merged it :)